### PR TITLE
[6/n][torch/elastic][upstream] Move torchelastic/distributed/api to torch/distributed/elastic/launchers/api

### PR DESCRIPTION
**Pull request by aivanou**
Summary: Move torchelastic/distributed/api to torch/distributed/elastic/launchers/api

Test Plan:
buck test mode/dev-nosan //pytorch/elastic/torchelastic/...
    buck test mode/dev-nosan //caffe2/test/distributed/elastic/agent/server/test/...

Differential Revision: D27352777

### CODE REVIEW

**Comment by cbalioglu** [comment]
Looks good to me. Do you mind checking the mypy error though?

File: test/distributed/elastic/launchers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env python3
```
Comment (mrshenli) [inline test/distributed/elastic/launchers/__init__.py:1]: Would I be correct if I assume, there will be followup efforts to consolidate with `https://github.com/pytorch/pytorch/blob/master/torch/distributed/launch.py`?

File: test/distributed/elastic/launchers/api_test.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
+#!/usr/bin/env python3
```
Comment (mrshenli) [inline test/distributed/elastic/launchers/api_test.py:1]: do we plan to plug `api_test.py` to `test/run_test.py` in the same as other distributed tests, or do we plan to use a separate test launching solution. If it's the former and if the tests are supposed to pass today, shall we enable them in this PR. If it's the later, we might need to make sure that we can still create an issue with title format "DISABLED testname.TestClass" to disable it in CI.

File: test/distributed/elastic/launchers/bin/test_script.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python3
```
Comment (mrshenli) [inline test/distributed/elastic/launchers/bin/test_script.py:1]: nit: existing tests in the distributed folder use the following convention: 1) file with a `test_` prefix are entry points of tests and is registered in `test/run_test.py`. And we can use things like `pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_func_name -vs` to run individual tests. 2) if some test implementation needs to shared across different test runs, we create a file with `_test` for implementation and import that in multiple `test_` prefix files, e.g., `rpc_test.py` is imported by both `test_tensorpipe_agent.py` and `test_process_group_agent.py`.

I am not sure if makes sense to follow the same convention here, but might be good to do so if it's not too much efforts.

**Pull request closed without merging**